### PR TITLE
Implement

### PR DESCRIPTION
**Pull request by danobi**
This is the V0 implementation of libbpf-rs.

### CODE REVIEW

File: libbpf-rs/src/object.rs [hunk 1]
```diff
@@ -1,84 +1,353 @@
...
-        unimplemented!();
+    pub fn set_max_entries(&mut self, entries: u32) -> &mut Self {
+        self.attrs.max_entries = entries;
+        self
     }
 
-    pub fn set_initial_value(&mut self, _data: &[u8]) -> &mut Self {
-        unimplemented!();
+    pub fn set_initial_value(&mut self, data: &[u8]) -> &mut Self {
+        self.initial_val = Some(data.to_vec());
```
Comment (anakryiko) [inline libbpf-rs/src/object.rs:292]: why not just calling `libbpf_sys::bpf_map__set_initial_value` here?

File: libbpf-rs/src/object.rs [hunk 2]
```diff
@@ -1,84 +1,353 @@
...
+                return Err(Error::System(-ret));
+            }
+        }
+
+        // Set name here because self.name could be memmoved during object construction
+        let name_cstring = util::str_to_cstring(self.name.as_str())?;
+        self.attrs.name = name_cstring.as_ptr();
+
+        // libbpf_sys::bpf_object__load() already created the map for us. So we just
+        // need to get the FD out.
```
Comment (anakryiko) [inline libbpf-rs/src/object.rs:332]: libbpf has low-level API to create maps independently from bpf_object, which will return FD. But as part of bpf_object, you can't just pick subset of BPF maps and create them. I'm not sure it's a good idea to create an illusion of existence of such API, if libbpf doesn't have it (and doesn't plan to have it).

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -12,12 +12,13 @@ jobs:
     steps:
     - uses: actions/checkout@v2
     - name: Install deps
-      run: sudo apt-get install -y clang-9
+      run: sudo apt-get install -y clang-9 libelf-dev
```
[the remark this answers is not in the document]
Comment (danobi) [reply .github/workflows/rust.yml:15]: Yeah but it's just for smoke testing. We don't try to run the progs. Reason it's 9 is b/c ubuntu doesn't have 10/11 yet.

File: libbpf-rs/src/object.rs [hunk 4]
```diff
@@ -1,84 +1,353 @@
...
+        };
+
+        let opts = self.opts(name_ptr);
+
+        let obj = unsafe { libbpf_sys::bpf_object__open_file(path_ptr, &opts) };
+        if obj.is_null() {
+            return Err(Error::Internal("Could not create bpf_object".to_string()));
+        }
+
+        let mut load_attr = libbpf_sys::bpf_object_load_attr {
```
[the remark this answers is not in the document]
Comment (danobi) [reply libbpf-rs/src/object.rs:107]: Ok, will think again

File: libbpf-rs/src/object.rs [hunk 5]
```diff
@@ -1,84 +1,353 @@
...
 
-    pub fn name(&self) -> &str {
-        unimplemented!();
-    }
+/// Represents a BPF object file. An object may contain zero or more
+/// [`Program`]s and [`Map`]s.
+pub struct Object {
+    ptr: *mut libbpf_sys::bpf_object,
+    maps: HashMap<String, MapBuilder>,
+    progs: HashMap<String, ProgramBuilder>,
```
[the remark this answers is not in the document]
Comment (danobi) [reply libbpf-rs/src/object.rs:173]: Yeah it's a little weird. Will think about this

**Comment by danobi** [comment]
> Map::load and Program::load pretend that they actually load and then own Map and Program. That's not really how libbpf operates, so I don't think that's a good idea to deviate so much in terms of API. It will just cause confusion down the line.

> You seem to keep pushing in the direction of having this control over individual maps/programs, what's the use case you are targeting?

Initially I thought it'd be possible (from looking at libbpf api) and feel more natural. Now that I know it doesn't quite work like that, I think it's not worth the effort to deviate from libbpf.

> Map::close/Program::clone -- what's the use case? This FD cloning looks fish and drop closing FD also looks wrong. In libbpf, bpf_object closes programs/maps on bpf_object__close(). You can bypass it with low-level API, but that's not encouraged way.

Yeah, will get rid of it.

I think I'll fix up the inline comments in this PR and then merge. I'll address the high level comments in a follow up. The stack of commits is annoying to work with.

File: libbpf-rs/src/object.rs [hunk 6]
```diff
@@ -1,84 +1,353 @@
...
+        let mut attrs = libbpf_sys::bpf_create_map_attr::default();
+        attrs.map_type = def.type_;
+        attrs.key_size = def.key_size;
+        attrs.value_size = def.value_size;
+        attrs.max_entries = def.max_entries;
+        attrs.map_flags = def.map_flags;
+
+        if attrs.map_type == (MapType::PerfEventArray as u32) && attrs.max_entries == 0 {
+            attrs.max_entries = unsafe { libbpf_sys::libbpf_num_possible_cpus() as u32 };
+        }
```
[the remark this answers is not in the document]
Comment (danobi) [reply libbpf-rs/src/object.rs:265]: I think I did this before when I was creating the map in libbpf-rs. After I realized libbpf creates the map I think this is leftover logic. I'll clean it up

File: libbpf-rs/src/object.rs (hunk 1 above)
Comment (danobi) [reply libbpf-rs/src/object.rs:292]: Yeah, I guess that'll work b/c the final `load()` (will probably rename to `build()`) acts as a barrier so users won't ever get access to a partially built map

Comment (danobi) [reply libbpf-rs/src/object.rs:292]: Hmm, actually the reason it's like that is b/c bpf_map__set_initial_value() can fail and it'd be weird to have a builder option be fallible

**Comment by danobi** [comment]
Todo in next PR:
* separate out object open and object load
* no independent map & prog loading
* make map & prog objects only accessible by reference

**Pull request merged**